TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Service/NodeTreeService.php

```diff
@@ -0,0 +1,52 @@
+<?php
+namespace TYPO3\TYPO3CR\Service;
+
+/*
+ * This file is part of the TYPO3.TYPO3CR package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use TYPO3\Flow\Annotations as Flow;
+use TYPO3\TYPO3CR\Domain\Factory\NodeFactory;
+use TYPO3\TYPO3CR\Domain\Model\NodeInterface;
+
+/**
+ * The node tree service
+ * @Flow\Scope("singleton")
+ */
+class NodeTreeService
+{
+
+    /**
+     * @Flow\Inject
+     * @var NodeFactory
+     */
+    protected $nodeFactory;
+
+
+    /**
+     * @param NodeInterface $node
+     * @param callable $action
+     * @return void
+     */
+    public function traverseTree(NodeInterface $node, callable $action)
+    {
+        $continue = $action($node);
+        if ($continue !== false && $node->hasChildNodes()) {
+            foreach ($node->getChildNodes() as $childNode) {
+                $this->traverseTree($childNode, $action);
+            }
+        }
+        $this->nodeFactory->removeNodeFromCache($node);
```

Comment:
Well, that's basically how you implement depth-first tree traversal: Build up a stack of traversed nodes and clean it up afterwards. The node caches (could be only one, perhaps?) in this case exactly play the role of that stack, so removal of items has to be exposed imho.